ZAPD/ZActorList.cpp

```diff
@@ -0,0 +1,194 @@
+#include "ZActorList.h"
+
+#include "Globals.h"
+#include "Utils/BitConverter.h"
+#include "WarningHandler.h"
+#include "ZFile.h"
+#include "ZRoom/ZNames.h"
+
+REGISTER_ZFILENODE(ActorList, ZActorList);
+
+ZActorList::ZActorList(ZFile* nParent) : ZResource(nParent)
+{
+	RegisterRequiredAttribute("Count");
+}
+
+void ZActorList::ExtractFromBinary(uint32_t nRawDataIndex, uint8_t nNumActors)
+{
+	rawDataIndex = nRawDataIndex;
+	numActors = nNumActors;
+
+	// Don't parse raw data of external files
+	if (parent->GetMode() == ZFileMode::ExternalFile)
+		return;
+
+	ParseRawData();
+}
+
+void ZActorList::ParseXML(tinyxml2::XMLElement* reader)
+{
+	ZResource::ParseXML(reader);
+
+	numActors = StringHelper::StrToL(registeredAttributes.at("Count").value);
+
+	if (numActors < 1)
+	{
+		HANDLE_ERROR_RESOURCE(
+			WarningType::InvalidAttributeValue, parent, this, rawDataIndex,
+			StringHelper::Sprintf("invalid value '%d' found for 'NumPaths' attribute", numActors),
+			"Should be at least '1'");
+	}
+}
+
+void ZActorList::ParseRawData()
+{
+	ZResource::ParseRawData();
+
+	offset_t currentPtr = rawDataIndex;
+	size_t largestlength = 0;
+
+	for (size_t i = 0; i < numActors; i++)
+	{
+		ActorSpawnEntry entry(parent->GetRawData(), currentPtr);
+
+		currentPtr += entry.GetRawDataSize();
+		actors.push_back(entry);
+
+		size_t actorNameLength = ZNames::GetActorName(entry.GetActorId()).size();
+		if (actorNameLength > largestlength)
+			largestlength = actorNameLength;
+	}
+
+	for (auto& entry : actors)
+	{
+		entry.SetLargestActorName(largestlength);
+	}
+}
+
+Declaration* ZActorList::DeclareVar(const std::string& prefix, const std::string& bodyStr)
+{
+	std::string auxName = name;
+
+	if (name == "")
+		auxName = GetDefaultName(prefix);
+
+	Declaration* decl =
+		parent->AddDeclarationArray(rawDataIndex, GetDeclarationAlignment(), GetRawDataSize(),
+	                                GetSourceTypeName(), name, GetActorListArraySize(), bodyStr);
+	decl->staticConf = staticConf;
+
+	return decl;
+}
+
+std::string ZActorList::GetBodySourceCode() const
+{
+	std::string declaration;
+
+	size_t index = 0;
+	for (auto& entry : actors)
+	{
+		declaration += StringHelper::Sprintf("\t{ %s },", entry.GetBodySourceCode().c_str());
+
+		if (index < actors.size() - 1)
+			declaration += "\n";
+
+		index++;
+	}
+
+	return declaration;
+}
+
+std::string ZActorList::GetSourceTypeName() const
+{
+	return actors.front().GetSourceTypeName();
```

Comment:
Might it make sense to set `GetSourceTypeName()` here to a static function accessed like `return ActorSpawnEntry::GetSourceTypeName()`? It's not an override, and it would ensure that in a scenario where the actor list is somehow empty, it wouldn't try to access a null pointer.